core/src/main/java/com/linecorp/armeria/internal/Http2KeepAliveHandler.java

```diff
@@ -0,0 +1,237 @@
+        this.pingTimeoutMillis = pingTimeoutMillis;
+        this.http2Connection = requireNonNull(http2Connection, "http2Connection");
+        this.sendPingsOnNoActiveStreams = sendPingsOnNoActiveStreams;
+    }
+
+    public void onChannelIdle(ChannelHandlerContext ctx, IdleStateEvent event) {
+        logger.debug("{} {} event triggered on channel.", channel, event);
+
+        if (!canSendPing()) {
+            // The default behaviour is to shutdown the channel on idle timeout if not HTTP 2.0 conn.
+            // So preserving the behaviour.
+            closeChannelAndLog();
+            return;
+        }
+
+        // Only interested in ALL_IDLE event and when Http2KeepAliveHandler is ready.
+        // Http2KeepAliveHandler will not be ready because IdleStateEvent events are emitted
+        // more faster than ping acks are received.
+        if (state != State.IDLE || event.state() != IdleState.ALL_IDLE) {
+            return;
+        }
+
+        writePing(ctx);
+    }
+
+    private boolean canSendPing() {
+        if (http2Connection.numActiveStreams() == 0) {
+            return sendPingsOnNoActiveStreams;
+        } else {
+            return true;
+        }
+    }
+
+    private void writePing(ChannelHandlerContext ctx) {
+        lastPingPayload = random.nextLong();
+        state = State.PING_SCHEDULED;
+        pingWriteFuture = frameWriter.writePing(ctx, false, lastPingPayload, ctx.newPromise())
+                .addListener(pingWriteListener);
+        ctx.flush();
+    }
+
+    /**
+     * Callback for when channel is in-active to cleans up resources.
+     */
+    public void onChannelInactive() {
+        state = State.SHUTDOWN;
+        if (shutdownFuture != null) {
+            shutdownFuture.cancel(false);
+            shutdownFuture = null;
+        }
+        if (pingWriteFuture != null) {
+            pingWriteFuture.cancel(false);
+            pingWriteFuture = null;
+        }
+    }
+
+    public void onPingAck(long data) throws Http2Exception {
+        final long elapsed = stopwatch.elapsed(TimeUnit.NANOSECONDS);
+        if (!isGoodPingAck(data)) {
+            return;
+        }
+
+        if (shutdownFuture != null) {
+            final boolean isCancelled = shutdownFuture.cancel(false);
+            if (!isCancelled) {
+                logger.debug("{} shutdownFuture cannot be cancelled because of late PING ACK", channel);
+            }
+        }
+        logger.debug("{} Received PING(ACK=1) in {} ns", channel, elapsed);
+        state = State.IDLE;
+    }
+
+    private boolean isGoodPingAck(long data) {
+        if (state != State.PENDING_PING_ACK) {
+            logger.warn("Unexpected PING(ACK=1, DATA={}) received", data);
+            return false;
+        }
+        if (lastPingPayload != data) {
+            logger.warn("Unexpected PING(ACK=1, DATA={}) received, " +
+                    "but expecting PING(ACK=1, DATA={})", data, lastPingPayload);
```

Comment:
How about logging the current channel as well?

    "{} Unexpected PING(ACK=1, DATA={}) received", channel, data
    "{} Unexpected PING(ACK=1, DATA={}) received ....